### PR TITLE
ERC20->NFT on seaport

### PR DESCRIPTION
**Pull request by hensha256**

### CODE REVIEW

File: src/entities/protocols/seaport.ts [hunk 1]
```diff
@@ -4,6 +4,9 @@ import { Interface } from '@ethersproject/abi'
 import { BuyItem, Market, NFTTrade, TokenType } from '../NFTTrade'
 import { TradeConfig } from '../Command'
 import { RoutePlanner, CommandType } from '../../utils/routerCommands'
+import { encodePermit, Permit2Permit, Permit2TransferFrom } from '../../utils/permit2'
+import { CONDUIT_SPENDER_ID, ETH_ADDRESS } from '../../utils/constants'
+import invariant from 'tiny-invariant'
```
Comment (marktoda) [inline src/entities/protocols/seaport.ts:9]: super nit: external imports before local

File: src/utils/routerCommands.ts [hunk 2]
```diff
@@ -57,36 +58,62 @@ const REVERTIBLE_COMMANDS = new Set<CommandType>([
   CommandType.FOUNDATION,
   CommandType.SUDOSWAP,
   CommandType.NFT20,
+  CommandType.EXECUTE_SUB_PLAN,
   CommandType.CRYPTOPUNKS,
+  CommandType.ELEMENT_MARKET,
 ])
 
+const PERMIT_STRUCT =
+  '((address token,uint160 amount,uint48 expiration,uint48 nonce) details, address spender, uint256 sigDeadline)'
```
Comment (marktoda) [inline src/utils/routerCommands.ts:67]: sometimes spaces sometimes not spaces after commas?

Comment (marktoda) [reply src/utils/routerCommands.ts:67]: maybe doesnt matter for ethers abiencoder but for consistency

File: src/entities/protocols/seaport.ts [hunk 3]
```diff
@@ -14,6 +17,9 @@ export type SeaportData = {
   items: Order[]
   recipient: string // address
   version: SeaportVersion
+  inputTokenApproval?: string
+  inputTokenPermit?: Permit2Permit
+  inputTokenTransfer?: Permit2TransferFrom
```
Comment (ewilz) [inline src/entities/protocols/seaport.ts:22]: I'm wondering if these merit code comments.... and how they work together.
Is there a situation where `inputTokenApproval` would be false, but the other two fields are filled in?

Comment (zhongeric) [reply src/entities/protocols/seaport.ts:22]: We're leaving approvals hanging, but yeah I agree it would be good to have a detailed comment here specifically about the `inputTokenApproval` as well as it's not intuitive that it stands for the approval of the opensea conduit by the Universal router for the ERC20. Someone else might think that it's for approving their tokens for universal router

File: src/entities/protocols/seaport.ts [hunk 4]
```diff
@@ -183,21 +216,21 @@ export class SeaportTrade extends NFTTrade<SeaportData> {
...
+    return { advancedOrder }
   }
 
-  private calculateValue(considerations: ConsiderationItem[]): BigNumber {
+  private calculateValue(considerations: ConsiderationItem[], token: string): BigNumber {
     return considerations.reduce(
-      (amt: BigNumber, consideration: ConsiderationItem) => amt.add(consideration.startAmount),
+      (amt: BigNumber, consideration: ConsiderationItem) =>
+        consideration.token == token ? amt.add(consideration.startAmount) : amt,
       BigNumber.from(0)
```
Comment (zhongeric) [inline src/entities/protocols/seaport.ts:234]: Is it possible for an order to have consideration items in different token denominations? I dont think so but good to make sure

File: src/utils/constants.ts [hunk 5]
```diff
@@ -58,6 +58,11 @@ export const PERMIT2_ADDRESS = '0x000000000022D473030F116dDEE9F6B43aC78BA3'
 
 export const CONTRACT_BALANCE = BigNumber.from(2).pow(255)
 export const ETH_ADDRESS = '0x0000000000000000000000000000000000000000'
+export const MAX_UINT256 = BigNumber.from(2).pow(256).sub(1)
+export const MAX_UINT160 = BigNumber.from(2).pow(160).sub(1)
 
 export const SENDER_AS_RECIPIENT = '0x0000000000000000000000000000000000000001'
 export const ROUTER_AS_RECIPIENT = '0x0000000000000000000000000000000000000002'
+
+export const CONDUIT_SPENDER_ID = 0
```
Comment (zhongeric) [inline src/utils/constants.ts:67]: `OPENSEA_CONDUIT_SEPNDER_ID`

File: test/forge/SwapNFTCallParameters.t.sol [hunk 6]
```diff
@@ -118,8 +123,62 @@ contract swapNFTCallParametersTest is Test, Interop, DeployRouter {
...
+        ERC721 token = ERC721(0x5180db8F5c931aaE63c74266b211F580155ecac8);
+        assertEq(token.balanceOf(RECIPIENT), 0);
+
+        (success,) = address(router).call{value: params.value}(params.data);
+        require(success, "call failed");
+        assertLt(WETH.balanceOf(from), balance);
+        assertEq(token.balanceOf(RECIPIENT), 1);
+    }
+
+    function testSeaportV1_4BuyItemsETH() public {
```
Comment (zhongeric) [inline test/forge/SwapNFTCallParameters.t.sol:180]: Should probably add the tests above with permit + approval for seaport 1.4 to ensure that the approval on the opensea conduit is working for 1.4 as well.

File: test/orders/seaport.ts [hunk 7]
```diff
@@ -109,13 +109,67 @@ export const seaportData2Covens: SeaportData = {
...
+        salt: '0x360c6ebe00000000000000000000000000000000000000006e312a018aa5d380',
+        conduitKey: '0x0000007b02230091a7ed01230072f7006a004d60a8d4e71d599b8104250f0000',
+        totalOriginalConsiderationItems: 3,
+      },
+      signature:
+        '0x595bd8731a7f40efcf819b872f4388ee836d884bbfc0b6a00131335c8923025245a78fabf360586ae157f34201bb5950a4858200bf69487178629dff43ab51c21b',
+    },
+  ],
+  recipient: TEST_RECIPIENT_ADDRESS,
+  version: SeaportVersion.V1_1,
```
Comment (zhongeric) [inline test/orders/seaport.ts:166]: Does the Opensea API return this version key? iirc it returns a `protocol_address`

File: package.json [hunk 8]
```diff
@@ -47,7 +47,7 @@
     "typescript": "^4.3.3"
   },
   "dependencies": {
-    "@uniswap/universal-router": "1.3.1",
+    "@uniswap/universal-router": "1.4.0-beta.0",
```
Comment (ewilz) [inline package.json:50]: just DM'd. I think all unviersal-router upgrades should be in their own commit titled the version name. It makes looking at the history a lot easier and know when certain commits were released at a glance

File: src/entities/protocols/seaport.ts [hunk 9]
```diff
@@ -98,9 +105,35 @@ export class SeaportTrade extends NFTTrade<SeaportData> {
...
+      }
+
+      if (!!order.inputTokenTransfer) {
+        planner.addCommand(CommandType.PERMIT2_TRANSFER_FROM, [
+          order.inputTokenTransfer.token,
+          order.inputTokenTransfer.routerAddress,
+          this.getTotalOrderPrice(order, order.inputTokenTransfer.token).toString(),
+        ])
+      }
+
```
[the remark this answers is not in the document]
Comment (ewilz) [reply src/entities/protocols/seaport.ts:133]: ++ and maybe there can be a type PermitOptions that encapsulates the 3 new params, with some good code comments for ppl like me 😜

File: test/orders/seaport.ts (hunk 7 above)
Comment (ewilz) [reply test/orders/seaport.ts:166]: o00o0o nice, I think accepting the address would be easier then. I initially tried to match api calls as much as possible

File: test/forge/SwapNFTCallParameters.t.sol [hunk 10]
```diff
@@ -12,6 +13,10 @@ import {ICryptopunksMarket} from "./utils/ICryptopunksMarket.sol";
 contract swapNFTCallParametersTest is Test, Interop, DeployRouter {
     using stdJson for string;
 
+    ERC20 private constant WETH = ERC20(0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2);
+    bytes constant APPROVE_WETH_DATA =
```
[the remark this answers is not in the document]
Comment (hensha256) [reply test/forge/SwapNFTCallParameters.t.sol:17]: I tried this but for some reason `forge` was then interpreting the output as a string not bytes. Very strange 🤪

[the remark this answers is not in the document]
Comment (hensha256) [reply test/forge/SwapNFTCallParameters.t.sol:17]: Will add a comment to explain it though

File: src/entities/protocols/seaport.ts (hunk 4 above)
Comment (hensha256) [reply src/entities/protocols/seaport.ts:224]: while i dont think that 1 consideration can, we actually sometimes submit multiple considerations in the same seaport command. So that means we could well have 2 considerations with different input currencies. Have updated to deal with this

File: test/forge/SwapNFTCallParameters.t.sol (hunk 6 above)
Comment (hensha256) [reply test/forge/SwapNFTCallParameters.t.sol:181]: good idea! will try to find some v1.4 tests once ive got element market in 👯

File: src/entities/protocols/seaport.ts [hunk 11]
```diff
@@ -1,19 +1,24 @@
...
-  version: SeaportVersion
+  protocolAddress: string
+  inputTokenProcessing?: InputTokenProcessing[]
+}
+
+export type InputTokenProcessing = {
+  token: string
+  permit2Permit?: Permit2Permit
+  protocolApproval: boolean
+  permit2TransferFrom: boolean
```
Comment (zhongeric) [inline src/entities/protocols/seaport.ts:21]: nit: place `protocolApproval` above `permit2Permit` to follow order of arguments in `encodeHandleInputTokens`

File: test/orders/seaportV1_4.ts [hunk 12]
```diff
@@ -56,10 +56,10 @@ export const seaportV1_4Data: SeaportData = {
     },
   ],
   recipient: TEST_RECIPIENT_ADDRESS,
-  version: SeaportVersion.V1_4,
+  protocolAddress: '0x00000000000001ad428e4906aE43D8F9852d0dD6',
```
Comment (zhongeric) [inline test/orders/seaportV1_4.ts:59]: ^ same as above

**Review by zhongeric — APPROVED** [review]

File: package.json (hunk 8 above)
Comment (ewilz) [reply package.json:52]: not addressed?

File: src/entities/protocols/seaport.ts [hunk 13]
```diff
@@ -98,9 +103,22 @@ export class SeaportTrade extends NFTTrade<SeaportData> {
           100, // TODO: look into making this a better number
         ])
       }
+
+      if (!!order.inputTokenProcessing) {
+        for (const inputToken of order.inputTokenProcessing)
+          encodeHandleInputTokens(
+            planner,
+            inputToken.protocolApproval ? { token: inputToken.token, protocol: order.protocolAddress } : undefined,
```
Comment (ewilz) [inline src/entities/protocols/seaport.ts:111]: are we supposed to approve to the protocol address or the conduit?

**Review by marktoda — APPROVED** [review]

File: src/entities/protocols/unwrapWETH.ts [hunk 14]
```diff
@@ -26,8 +23,10 @@ export class UnwrapWETH implements Command {
   }
 
   encode(planner: RoutePlanner, _: TradeConfig): void {
-    if (!!this.permit2Data) encodePermit(planner, this.permit2Data)
-    planner.addCommand(CommandType.PERMIT2_TRANSFER_FROM, [this.wethAddress, this.routerAddress, this.amount])
+    encodeHandleInputTokens(planner, undefined, this.permit2Data, {
```
Comment (ewilz) [inline src/entities/protocols/unwrapWETH.ts:26]: I personally tend to be pretty averse to functions that can have random `undefined`'s sprinkled in its function parameters.
Personal preference, but I think we also tend to do this in other sdk's maybe there could be options when not everything is required:

```
type InputTokenOptions {
   approval?: ApproveProtocol,
   permit2Data? Permit2Permit,
   permit2TransferFrom?: Permit2TransferFromOptions
}

encodeHandleInputTokens(
  planner, 
  { 
     permit2Data: this.permit2Data,
     permit2TransferFrom: {
         token: this.wethAddress,
         amount: this.amount.toString()
  }

)
```

Comment (marktoda) [reply src/entities/protocols/unwrapWETH.ts:26]: ah i prefer this too fwiw

File: test/orders/seaport.ts [hunk 15]
```diff
@@ -106,16 +106,70 @@ export const seaportData2Covens: SeaportData = {
     },
   ],
   recipient: TEST_RECIPIENT_ADDRESS,
-  version: SeaportVersion.V1_1,
+  protocolAddress: '0x00000000006c3852cbEf3e08E8dF289169EdE581',
```
[the remark this answers is not in the document]
Comment (hensha256) [reply test/orders/seaport.ts:109]: If I put it _inside_ the items array then I have to have a check that the `n` items inside the array all have identical `protocolAddress` - so it felt simpler to just have it appear once and not have to do that check. But could move it. Thoughts?

File: test/utils/permit2.ts [hunk 16]
```diff
@@ -37,7 +37,7 @@ export function toInputPermit(signature: string, permit: PermitSingle): Permit2P
 
 export function makePermit(
   token: string,
-  amount: string = ethers.constants.MaxUint256.toString(),
+  amount: string = MAX_UINT160.toString(),
```
[the remark this answers is not in the document]
Comment (hensha256) [reply test/utils/permit2.ts:40]: [Thats the maximum allowed by Permit2](https://github.com/Uniswap/permit2-sdk/blob/84d24d6d58a2457b55d70951c78ed2f862b18f39/src/constants.ts#L11) - it was wrong before

File: test/forge/SwapNFTCallParameters.t.sol (hunk 10 above)
[the remark this answers is not in the document]
Comment (ewilz) [reply test/forge/SwapNFTCallParameters.t.sol:18]: can you wrap bytes("arbitrarystring") ?

File: test/orders/seaport.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (zhongeric) [reply test/orders/seaport.ts:109]: It depends on where it is in the api response object returned by the OS api, I don't have an api key so can't check but we should verify
